Description: Do not depend on zlib or RLE compression being available.
 The Debian OpenSSL package is built without support for zlib
 compression since version 1.0.1e-5, and OpenSSL does not support
 RLE compression in all versions.
Forwarded: not-yet
Author: Peter Pentchev <roam@ringlet.net>
Last-Update: 2015-06-13

--- a/src/ssl.c
+++ b/src/ssl.c
@@ -115,13 +115,35 @@
     }
 
     /* also insert one of obsolete (ZLIB/RLE) algorithms */
-    if(global->compression==COMP_ZLIB) {
-        /* 224 - within the private range (193 to 255) */
-        SSL_COMP_add_compression_method(0xe0, COMP_zlib());
-    } else if(global->compression==COMP_RLE) {
-        /* 225 - within the private range (193 to 255) */
-        SSL_COMP_add_compression_method(0xe1, COMP_rle());
+    int id;
+    COMP_METHOD *comp;
+    switch (global->compression)
+    {
+#ifdef COMP_ZLIB
+        case COMP_ZLIB:
+            /* 224 - within the private range (193 to 255) */
+            id = 0xe0;
+            comp = COMP_zlib();
+            break;
+#endif
+
+#ifdef COMP_RLE
+        case COMP_RLE:
+            /* 225 - within the private range (193 to 255) */
+            id = 0xe1;
+            comp = COMP_rle();
+            break;
+#endif
+
+        default:
+            s_log(LOG_ERR, "Unknown or unsupported compression method specified");
+            return 1;
+    }
+    if (comp == NULL) {
+        s_log(LOG_ERR, "Unsupported compression method");
+        return 1;
     }
+    SSL_COMP_add_compression_method(id, comp);
     s_log(LOG_INFO, "Compression enabled: %d method(s)",
         sk_SSL_COMP_num(methods));
     return 0; /* success */
--- a/src/options.c
+++ b/src/options.c
@@ -439,9 +439,9 @@
         if(strcasecmp(opt, "compression"))
             break;
         if(SSLeay()>=0x00908051L && !strcasecmp(arg, "deflate"))
-            new_global_options.compression=COMP_DEFLATE;
+            return "deflate compression is not really implemented in OpenSSL";
         else if(!strcasecmp(arg, "zlib"))
-            new_global_options.compression=COMP_ZLIB;
+            return "zlib compression is disabled in the Debian package of OpenSSL";
         else if(!strcasecmp(arg, "rle"))
             new_global_options.compression=COMP_RLE;
         else
--- a/configure.ac
+++ b/configure.ac
@@ -179,10 +179,9 @@
 AC_SEARCH_LIBS([yp_get_default_domain], [nsl])
 AC_SEARCH_LIBS([socket], [socket])
 AC_SEARCH_LIBS([openpty], [util])
-# Checks for dynamic loader and zlib needed by OpenSSL
+# Checks for dynamic loader needed by OpenSSL
 AC_SEARCH_LIBS([dlopen], [dl])
 AC_SEARCH_LIBS([shl_load], [dld])
-AC_SEARCH_LIBS([inflateEnd], [z])
 
 # Add BeOS libraries
 if test "$host_os" = "beos"; then
